driver-core/src/main/com/mongodb/ClientBulkWriteException.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.mongodb;
+
+import com.mongodb.bulk.WriteConcernError;
+import com.mongodb.client.model.bulk.ClientWriteModel;
+import com.mongodb.client.result.bulk.ClientBulkWriteResult;
+import com.mongodb.lang.Nullable;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.mongodb.assertions.Assertions.isTrueArgument;
+import static java.util.Collections.emptyList;
+import static java.util.Collections.emptyMap;
+import static java.util.Collections.unmodifiableList;
+import static java.util.Collections.unmodifiableMap;
+import static java.util.Optional.ofNullable;
+
+/**
+ * The result of an unsuccessful or partially unsuccessful client-level bulk write operation.
+ *
+ * @see ClientBulkWriteResult
+ * @since 5.3
+ * @serial exclude
+ */
+public final class ClientBulkWriteException extends MongoServerException {
+    private static final long serialVersionUID = 1;
+
+    @Nullable
+    private final MongoException error;
+    private final List<WriteConcernError> writeConcernErrors;
+    private final Map<Long, WriteError> writeErrors;
+    @Nullable
+    private final ClientBulkWriteResult partialResult;
+
+    private ClientBulkWriteException(
+            final String message,
+            @Nullable final MongoException error,
+            @Nullable final List<WriteConcernError> writeConcernErrors,
+            @Nullable final Map<Long, WriteError> writeErrors,
+            @Nullable final ClientBulkWriteResult partialResult,
+            final ServerAddress serverAddress) {
+        super(message, serverAddress);
+        // BULK-TODO Should ClientBulkWriteException.getCode be the same as error.getCode,
+        // and getErrorLabels/hasErrorLabel contain the same labels as error.getErrorLabels?
```

Comment:
Maybe reviewers have thoughts on this comment. My understanding is that the existing bulk write API does not wrap the top-level errors in `BulkWriteException`, but the new API wraps the top-level errors in `ClientBulkWriteException`. Given that the driver uses error code and labels to implement some of its functionality, the question seems reasonable. I don't currently have an answer.

Reply:
I will have to think about this more.  I'm not sure.

Reply:
These are all the error labels the driver knows about and uses: `TransientTransactionError`, `UnknownTransactionCommitResult`, `RetryableWriteError`, `NoWritesPerformed`.

This question splits into two:

1. What error codes/labels should be copied to `ClientBulkWriteException` from `ClientBulkWriteException.error` (the top-level error) for consumption by applications?
   - Simply copying some labels seems to be just incorrect, for example, `RetryableWriteError` - even if a specific `bulkWrite` command within a bulk operation can be retried, it does not mean that all previous commands can also be repeated when an application attempts to retry the whole operation. The same reasoning seems to apply to `NoWritesPerformed`. However, if no `bulkWrite` commands succeeded, then copying both of these labels seems safe.
   - I don't think we should copy any error codes for consumption by applications.
2. What error codes/labels should be copied to `ClientBulkWriteException` from `ClientBulkWriteException.error` (the top-level error) for consumption by the driver?
   - Given that `ClientBulkWriteException` will be created by `ClientBulkWriteOperation.execute`, only the following driver code will have a chance to handle labels and the code from this exception:
      - `MongoClusterImpl.OperationExecutorImpl.execute` - reads `TransientTransactionError`, `UnknownTransactionCommitResult`, and code 91.
      - `ClientSessionImpl.withTransaction` - reads `TransientTransactionError`, `UnknownTransactionCommitResult`.
   - It seems safe to copy the code if its 91.
   - It seems safe to copy the label `UnknownTransactionCommitResult`.
   - I am not sure about copying `TransientTransactionError`. @jyemin does this label always result in driver aborting the transaction?

Reply:
Seems safest not to copy any error codes / error labels to the top level Exception.

If a user needs to inspect them can't they use the `writeErrors` map?

Reply:
> Seems safest not to copy any error codes / error labels to the top level Exception.

I agree that we should not copy error codes/labels for consumption by applications. However, we may still decide to do that for consumption by the driver (and if we do so, the applications will be able to observe that).

I am OK with not copying anything, and instead updating the logic of `MongoClusterImpl.OperationExecutorImpl.execute`, `ClientSessionImpl.withTransaction`, and their reactive counterparts such that they look at `ClientBulkWriteException.getError`.

Reply:
> If a user needs to inspect them can't they use the writeErrors map?

I don't think that `ClientBulkWriteException.getWriteErrors` is relevant here. Only the code/labels from the top-level error seem to make sense to consider in the context of copying them to `ClientBulkWriteException`.

Reply:
Done in dbf9a26b8c636a2a91e0132697cfe50c7073fbac.